# COGS 108 - Data Checkpoint

# Names

- Asher Av
- Quoc-Zuy  Do
- Hector Gallo
- Jeremy Nurding
- Andres Villegas

<a id='research_question'></a>
# Research Question

*Fill in your research question here*

# Dataset(s)

*Fill in your dataset information here*

(Copy this information for each dataset)
- Dataset Name:
- Link to the dataset:
- Number of observations:

1-2 sentences describing each dataset. 

If you plan to use multiple datasets, add 1-2 sentences about how you plan to combine these datasets.


**Data Set Name: COVID-19 Data - US Counties from NYTimes**
- Link to Dataset: https://github.com/nytimes/covid-19-data
- Number of Obsevations: 2,170,941 
- <ins>Description of Dataset:</ins> This dataset is collated from  data across the U.S. by the New York Times and draws from the official reportings about the cumulative number of cases and deaths reported in each county and state across the U.S since the start of the COVID-19 pandemic. This dataset contains 6 columns of  data: date, county, state, fips, cases and  deaths. The FIPS column  crefers to a FIPS code, a geographic identifier that determines the location of the county the data was pulled from and makes it easy to associate with other datasets

**Data Set Name: Air Quality Across Countries in COVID-19**
- Link to Dataset: https://www.kaggle.com/aestheteaman01/air-quality-across-countries-in-covid19/version/3?select=USA.csv
- Number of Observations: 179,365
- <ins>Description of Dataset:</ins> This dataset shows the collected information of the air quality from 2020-2021 across countries including Brazil, Canada, France, Italy, India, and USA and their respective cities. The parameters that they use to show the air quality for each of the cities are the following: Carbon Monoxide (CO), Dew (dew), Humidity (humidity), Nitrogen Dioxide (NO2), Ozone (O3), Particulate Matter 10 (pm10, Particulate Matter 2.5 (pm25), Pressure (pressure), Sulphur Dioxide (SO2), Temperature, Wind Gusts, and Wind Speed.

**Data Set Name: Air Quality Statistics by County, 2020**
- Link to Dataset: https://www.epa.gov/air-trends/air-quality-cities-and-counties 
- Number of Observations: 1,144
- <ins>Description of Dataset:</ins> The data set “Air Quality Statistics by County, 2020” presents data of counties with highest levels of air pollutants across states in the United States.  This data set is organized alphabetically in rows by state, each county is assigned to its state accordingly and also arranged in alphabetical order.  There are a total of 1162 rows in this data set and out of the total of rows however 1144 rows contain observations of air pollutants for specific counties.  This dataset contains 13 columns from left to right as follows: State, County, County FIPS Code, 2010 population, CO, Pb, NO2 AM (ppb), NO2 hr (ppb), O3, PM10, PM2.5 Wtd AM, PM2.5 hr, and SO2.  These show the highest number of such pollutant in the given area.

# Setup

##### Import Modules

In [1]:
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt

In [2]:
# nytimes daily covid dataset
covid_df = pd.read_csv('https://raw.githubusercontent.com/nytimes/covid-19-data/master/us-counties.csv')

# EPA 2020 NO2 dataset
epa2020_df = pd.read_csv('https://raw.githubusercontent.com/asherbav/covid_pollution_files/main/epa2020.csv')

# EPA 2021 NO2 dataset 
epa2021_df = pd.read_csv('https://raw.githubusercontent.com/asherbav/covid_pollution_files/main/epa2021.csv')


# Data Cleaning

---
### COVID-19 Dataset Cleaning
The first thing that we want to do is take a look at the original datasets to see what they look like. We first take a look at the COVID-19 Dataset entitled: "COVID-19 Data - US Counties from NYTimes". 

In [3]:
covid_df

date      county       state     fips  cases  deaths
0        2020-01-21   Snohomish  Washington  53061.0      1     0.0
1        2020-01-22   Snohomish  Washington  53061.0      1     0.0
2        2020-01-23   Snohomish  Washington  53061.0      1     0.0
3        2020-01-24        Cook    Illinois  17031.0      1     0.0
4        2020-01-24   Snohomish  Washington  53061.0      1     0.0
...             ...         ...         ...      ...    ...     ...
2203466  2022-02-10  Sweetwater     Wyoming  56037.0  10790   119.0
2203467  2022-02-10       Teton     Wyoming  56039.0   9560    15.0
2203468  2022-02-10       Uinta     Wyoming  56041.0   5564    36.0
2203469  2022-02-10    Washakie     Wyoming  56043.0   2267    42.0
2203470  2022-02-10      Weston     Wyoming  56045.0   1523    17.0

[2203471 rows x 6 columns]

We then want to check if there are any null-values that we might want to get rid of  using isnull().sum. We see that there are a bunch of null values inside of the death's columns and the FIPS column. 
We also want to check the types of the columns to make sure that they are in the format that is desired for instance it seems that the dates were properly  converted to datetime objects, however the FIPS columns are floats instead of integers as expected.

In [4]:
print(covid_df.isnull().sum())
print('-----------------') 
covid_df.dtypes


date          0
county        0
state         0
fips      20470
cases         0
deaths    50429
dtype: int64
-----------------


date       object
county     object
state      object
fips      float64
cases       int64
deaths    float64
dtype: object

We however, will not be needing the deaths column for our analysis so we can drop the column entirely. We may want to keep the fips data however as creating a heatmaps of COVID-19 distribution across regions will allow us to analyze which regions are the most heavily impacted by the virus. This requires that we remove the entries where there are null values using dropna() from Pandas. 

We also want to change the type of the FIPS codes to integers so they are in the expected format.

In [5]:
covid_df = covid_df.drop(['deaths'], axis = 1)
covid_df = covid_df.dropna()
covid_df['fips'] = covid_df['fips'].astype(int)


At this point we are done cleaning the COVID-19 dataset. We run isnull() again to double check that all the NULL/NaN values have been removed from out dataset properly. We also display the dataset and can see that the FIPS codes are now integers  as expected.  At this point the COVID-19 dataset has been cleaned and is now in a state where it can be used in our project.

In [6]:
print(covid_df.isnull().sum())
print('-----------------') 
covid_df

date      0
county    0
state     0
fips      0
cases     0
dtype: int64
-----------------


date      county       state   fips  cases
0        2020-01-21   Snohomish  Washington  53061      1
1        2020-01-22   Snohomish  Washington  53061      1
2        2020-01-23   Snohomish  Washington  53061      1
3        2020-01-24        Cook    Illinois  17031      1
4        2020-01-24   Snohomish  Washington  53061      1
...             ...         ...         ...    ...    ...
2203466  2022-02-10  Sweetwater     Wyoming  56037  10790
2203467  2022-02-10       Teton     Wyoming  56039   9560
2203468  2022-02-10       Uinta     Wyoming  56041   5564
2203469  2022-02-10    Washakie     Wyoming  56043   2267
2203470  2022-02-10      Weston     Wyoming  56045   1523

[2183001 rows x 5 columns]

---
### Kaggle Air Quality Data Set Cleaning
We repeat the steps for the COVID dataset from above for the Kaggle dataset on Air Quality. We display the dataset to  get an idea of what it looks like, search for any NULL/NaN values, and then also check the types of all the objects to make sure if everything is in order.

In [7]:
#print(kagaq_df.isnull().sum())
print('-----------------') 
#print(kagaq_df.dtypes)
print('-----------------') 
#print(kagaq_df['Country'].value_counts())
#kagaq_df

-----------------
-----------------


From the above analysis we can see that there are no null values, and that data types of each column is as expected for this dataset. This dataset has already clean and we don't have to do much from here. However, this dataset has an extraneous column and that is the "Country" column. This particular dataset only had dataset for the United States of America, therefore this column does not provide any additional useful information we can safely drop this column.

In [8]:
#kagaq_df = kagaq_df.drop(['Country'], axis = 1)

Next we redisplay the data set and confirm that the column was removed properly.

In [9]:
#kagaq_df

---
### EPA 2020 Air Quality Data Set Cleaning
We repeat the steps for the COVID dataset from above for the epa datasets on Air Quality. We display the dataset to get an idea of what it looks like, search for any NULL/NaN values, and then also check the types of all the objects to make sure if everything is in order.

In [10]:
epa2020_df

Date Source   Site ID  POC  Daily Max 1-hour NO2 Concentration  \
0     01/01/2020    AQS  60730001    1                                  18   
1     01/02/2020    AQS  60730001    1                                  35   
2     01/03/2020    AQS  60730001    1                                  36   
3     01/04/2020    AQS  60730001    1                                  31   
4     01/05/2020    AQS  60730001    1                                  36   
...          ...    ...       ...  ...                                 ...   
2864  12/27/2020    AQS  60731026    1                                  11   
2865  12/28/2020    AQS  60731026    1                                  12   
2866  12/29/2020    AQS  60731026    1                                  29   
2867  12/30/2020    AQS  60731026    1                                  31   
2868  12/31/2020    AQS  60731026    1                                  33   

     UNITS  DAILY_AQI_VALUE                              Site Name  \
0      ppb               17                            Chula Vista   
1      ppb               33                            Chula Vista   
2      ppb               34                            Chula Vista   
3      ppb               29                            Chula Vista   
4      ppb               34                            Chula Vista   
...    ...              ...                                    ...   
2864   ppb               10  San Diego - Sherman Elementary School   
2865   ppb               11  San Diego - Sherman Elementary School   
2866   ppb               27  San Diego - Sherman Elementary School   
2867   ppb               29  San Diego - Sherman Elementary School   
2868   ppb               31  San Diego - Sherman Elementary School   

      DAILY_OBS_COUNT  PERCENT_COMPLETE  AQS_PARAMETER_CODE  \
0                  22              92.0               42602   
1                  22              92.0               42602   
2                  22              92.0               42602   
3                  22              92.0               42602   
4                  22              92.0               42602   
...               ...               ...                 ...   
2864               22              92.0               42602   
2865               22              92.0               42602   
2866               22              92.0               42602   
2867               20              83.0               42602   
2868               22              92.0               42602   

          AQS_PARAMETER_DESC  CBSA_CODE               CBSA_NAME  STATE_CODE  \
0     Nitrogen dioxide (NO2)      41740  San Diego-Carlsbad, CA           6   
1     Nitrogen dioxide (NO2)      41740  San Diego-Carlsbad, CA           6   
2     Nitrogen dioxide (NO2)      41740  San Diego-Carlsbad, CA           6   
3     Nitrogen dioxide (NO2)      41740  San Diego-Carlsbad, CA           6   
4     Nitrogen dioxide (NO2)      41740  San Diego-Carlsbad, CA           6   
...                      ...        ...                     ...         ...   
2864  Nitrogen dioxide (NO2)      41740  San Diego-Carlsbad, CA           6   
2865  Nitrogen dioxide (NO2)      41740  San Diego-Carlsbad, CA           6   
2866  Nitrogen dioxide (NO2)      41740  San Diego-Carlsbad, CA           6   
2867  Nitrogen dioxide (NO2)      41740  San Diego-Carlsbad, CA           6   
2868  Nitrogen dioxide (NO2)      41740  San Diego-Carlsbad, CA           6   

           STATE  COUNTY_CODE     COUNTY  SITE_LATITUDE  SITE_LONGITUDE  
0     California           73  San Diego      32.631242     -117.059088  
1     California           73  San Diego      32.631242     -117.059088  
2     California           73  San Diego      32.631242     -117.059088  
3     California           73  San Diego      32.631242     -117.059088  
4     California           73  San Diego      32.631242     -117.059088  
...          ...          ...        ...            ...             ...  
2864  Cali

In [11]:
print(epa2020_df.isnull().sum())
print('-----------------') 
epa2020_df.dtypes

Date                                  0
Source                                0
Site ID                               0
POC                                   0
Daily Max 1-hour NO2 Concentration    0
UNITS                                 0
DAILY_AQI_VALUE                       0
Site Name                             0
DAILY_OBS_COUNT                       0
PERCENT_COMPLETE                      0
AQS_PARAMETER_CODE                    0
AQS_PARAMETER_DESC                    0
CBSA_CODE                             0
CBSA_NAME                             0
STATE_CODE                            0
STATE                                 0
COUNTY_CODE                           0
COUNTY                                0
SITE_LATITUDE                         0
SITE_LONGITUDE                        0
dtype: int64
-----------------


Date                                   object
Source                                 object
Site ID                                 int64
POC                                     int64
Daily Max 1-hour NO2 Concentration      int64
UNITS                                  object
DAILY_AQI_VALUE                         int64
Site Name                              object
DAILY_OBS_COUNT                         int64
PERCENT_COMPLETE                      float64
AQS_PARAMETER_CODE                      int64
AQS_PARAMETER_DESC                     object
CBSA_CODE                               int64
CBSA_NAME                              object
STATE_CODE                              int64
STATE                                  object
COUNTY_CODE                             int64
COUNTY                                 object
SITE_LATITUDE                         float64
SITE_LONGITUDE                        float64
dtype: object

Since there are no null values present, we will not be excluding any rows. 

There appears to be many columns that we will not be using for our analysis. 

After looking at each column for their unique values, we noticed that we can safely drop the 'COUNTY', 'COUNTY_CODE','STATE','STATE_CODE','CBSA_NAME','CBSA_CODE','AQS_PARAMETER_DESC','AQS_PARAMETER_CODE','UNITS',and 'Source' columns as there is only a single value representing the entire dataset. 

The description of the dataset is as follows: The county and its associated code is 'San Diego', the CBSA is 'San Diego-Carlsbad, CA', the air quality parameter we're looking at is 'NO<sub>2</sub>', the unit is in ppb or parts per billion, and the source is 'AQS'. 

Furthermore the values 'SITE_LATITUDE','SITE_LONGITUDE','Site ID','POC','PERCENT_COMPLETE', and'DAILY_OBS_COUNT' are unnecessary for the analysis that we will be peforming. 

In short, the columns that we're interested in are the date, the NO2 levels, the AQI, and the particular sites. We will also be renaming multi-word labels. 

Lastly, we'll need to convert the Date to datetime objects. 

In [12]:
same_value_list = ['COUNTY', 'COUNTY_CODE','STATE','STATE_CODE','CBSA_NAME','CBSA_CODE','AQS_PARAMETER_DESC','AQS_PARAMETER_CODE','UNITS','Source']
not_used_list = ['SITE_LATITUDE','SITE_LONGITUDE','Site ID','POC','PERCENT_COMPLETE','DAILY_OBS_COUNT']
updated_labels = {'Daily Max 1-hour NO2 Concentration': 'NO2', 'DAILY_AQI_VALUE': 'AQI', 'Site Name': 'Site'}

In [13]:
# Drop columns that contain only a single value
epa2020_df = epa2020_df.drop(same_value_list, axis = 1)

# Drop columns that are not going to be used
epa2020_df = epa2020_df.drop(not_used_list, axis = 1)

# Rename columns
epa2020_df = epa2020_df.rename(columns=updated_labels)

# Convert date to datetime object
epa2020_df['Date'] = pd.to_datetime(epa2020_df['Date'])

In [14]:
epa2020_df

Date    datetime64[ns]
NO2              int64
AQI              int64
Site            object
dtype: object

Since the 2021 dataset from the EPA is of a similar format to the 2020 dataset, the same rationale that we used for cleaning the 2020 dataset will be applied. 

In [15]:
epa2021_df

Date Source   Site ID  POC  Daily Max 1-hour NO2 Concentration  \
0     01/01/2021    AQS  60730001    1                                  32   
1     01/02/2021    AQS  60730001    1                                  32   
2     01/03/2021    AQS  60730001    1                                  22   
3     01/04/2021    AQS  60730001    1                                  27   
4     01/05/2021    AQS  60730001    1                                  29   
...          ...    ...       ...  ...                                 ...   
2006  08/27/2021    AQS  60731026    2                                  12   
2007  08/28/2021    AQS  60731026    2                                   9   
2008  08/29/2021    AQS  60731026    2                                  10   
2009  08/30/2021    AQS  60731026    2                                  14   
2010  08/31/2021    AQS  60731026    2                                   4   

     UNITS  DAILY_AQI_VALUE                              Site Name  \
0      ppb               30                            Chula Vista   
1      ppb               30                            Chula Vista   
2      ppb               21                            Chula Vista   
3      ppb               25                            Chula Vista   
4      ppb               27                            Chula Vista   
...    ...              ...                                    ...   
2006   ppb               11  San Diego - Sherman Elementary School   
2007   ppb                8  San Diego - Sherman Elementary School   
2008   ppb                9  San Diego - Sherman Elementary School   
2009   ppb               13  San Diego - Sherman Elementary School   
2010   ppb                4  San Diego - Sherman Elementary School   

      DAILY_OBS_COUNT  PERCENT_COMPLETE  AQS_PARAMETER_CODE  \
0                  22              92.0               42602   
1                  22              92.0               42602   
2                  22              92.0               42602   
3                  22              92.0               42602   
4                  22              92.0               42602   
...               ...               ...                 ...   
2006               22              92.0               42602   
2007               22              92.0               42602   
2008               22              92.0               42602   
2009               22              92.0               42602   
2010               22              92.0               42602   

          AQS_PARAMETER_DESC  CBSA_CODE               CBSA_NAME  STATE_CODE  \
0     Nitrogen dioxide (NO2)      41740  San Diego-Carlsbad, CA           6   
1     Nitrogen dioxide (NO2)      41740  San Diego-Carlsbad, CA           6   
2     Nitrogen dioxide (NO2)      41740  San Diego-Carlsbad, CA           6   
3     Nitrogen dioxide (NO2)      41740  San Diego-Carlsbad, CA           6   
4     Nitrogen dioxide (NO2)      41740  San Diego-Carlsbad, CA           6   
...                      ...        ...                     ...         ...   
2006  Nitrogen dioxide (NO2)      41740  San Diego-Carlsbad, CA           6   
2007  Nitrogen dioxide (NO2)      41740  San Diego-Carlsbad, CA           6   
2008  Nitrogen dioxide (NO2)      41740  San Diego-Carlsbad, CA           6   
2009  Nitrogen dioxide (NO2)      41740  San Diego-Carlsbad, CA           6   
2010  Nitrogen dioxide (NO2)      41740  San Diego-Carlsbad, CA           6   

           STATE  COUNTY_CODE     COUNTY  SITE_LATITUDE  SITE_LONGITUDE  
0     California           73  San Diego      32.631242     -117.059088  
1     California           73  San Diego      32.631242     -117.059088  
2     California           73  San Diego      32.631242     -117.059088  
3     California           73  San Diego      32.631242     -117.059088  
4     California           73  San Diego      32.631242     -117.059088  
...          ...          ...        ...            ...             ...  
2006  Cali

In [16]:
print(epa2021_df.isnull().sum())
print('-----------------') 
epa2021_df.dtypes

Date                                  0
Source                                0
Site ID                               0
POC                                   0
Daily Max 1-hour NO2 Concentration    0
UNITS                                 0
DAILY_AQI_VALUE                       0
Site Name                             0
DAILY_OBS_COUNT                       0
PERCENT_COMPLETE                      0
AQS_PARAMETER_CODE                    0
AQS_PARAMETER_DESC                    0
CBSA_CODE                             0
CBSA_NAME                             0
STATE_CODE                            0
STATE                                 0
COUNTY_CODE                           0
COUNTY                                0
SITE_LATITUDE                         0
SITE_LONGITUDE                        0
dtype: int64
-----------------


Date                                   object
Source                                 object
Site ID                                 int64
POC                                     int64
Daily Max 1-hour NO2 Concentration      int64
UNITS                                  object
DAILY_AQI_VALUE                         int64
Site Name                              object
DAILY_OBS_COUNT                         int64
PERCENT_COMPLETE                      float64
AQS_PARAMETER_CODE                      int64
AQS_PARAMETER_DESC                     object
CBSA_CODE                               int64
CBSA_NAME                              object
STATE_CODE                              int64
STATE                                  object
COUNTY_CODE                             int64
COUNTY                                 object
SITE_LATITUDE                         float64
SITE_LONGITUDE                        float64
dtype: object

In [17]:
# Drop columns that contain only a single value
epa2021_df = epa2021_df.drop(same_value_list, axis = 1)

# Drop columns that are not going to be used
epa2021_df = epa2021_df.drop(not_used_list, axis = 1)

# Rename columns
epa2021_df = epa2021_df.rename(columns=updated_labels)

# Convert date to datetime object
epa2021_df['Date'] = pd.to_datetime(epa2021_df['Date'])

In [18]:
epa2021_df

Date  NO2  AQI                                   Site
0    2021-01-01   32   30                            Chula Vista
1    2021-01-02   32   30                            Chula Vista
2    2021-01-03   22   21                            Chula Vista
3    2021-01-04   27   25                            Chula Vista
4    2021-01-05   29   27                            Chula Vista
...         ...  ...  ...                                    ...
2006 2021-08-27   12   11  San Diego - Sherman Elementary School
2007 2021-08-28    9    8  San Diego - Sherman Elementary School
2008 2021-08-29   10    9  San Diego - Sherman Elementary School
2009 2021-08-30   14   13  San Diego - Sherman Elementary School
2010 2021-08-31    4    4  San Diego - Sherman Elementary School

[2011 rows x 4 columns]

Unique values from columns
 ---
 County codes all the same

 County all the same

 State all the same

 CBSA name all the same

 CBSA code all the same

 AQS_PARAMETER_DESC all the same

 AQS_PARAMETER_CODE all the same
 

 UNITS all the same in ppb(part per billion)

 Source all the same in AQS

 PERCENT_COMPLETE = [ 92., 100.,  83.,  88.,  75.,  79.,  96.]

 Site name = ['Chula Vista', 'Alpine', 'Camp Pendleton', 'Donovan',
       'Kearny Villa Rd.', 'San Diego -Rancho Carmel Drive',
       'El Cajon - Lexington Elementary School',
       'San Diego - Sherman Elementary School']

 DAILY_OBS_COUNT = [22, 24, 20, 21, 18, 19, 23]

 DAILY AQI VALUE = [17, 33, 34, 29, 38, 35, 28, 24, 30, 26, 21, 27, 36, 31, 25, 32, 15,
        23, 40, 16, 13, 39,  8, 22, 11, 12, 18,  7, 19, 20, 10, 14,  6,  4,
        9,  3,  5,  2, 41, 37, 42,  1, 53, 43, 55, 44, 45, 46, 47, 52, 48,
       49, 51, 50]

 Daily Max 1-hour NO2 Concentration = [18, 35, 36, 31, 40, 37, 30, 25, 32, 28, 22, 29, 38, 33, 26, 34, 16,
        24, 42, 17, 14, 41, 27,  8, 23, 12, 13, 19,  7, 20, 21, 11, 15,  9,
        6,  4, 10,  3,  5,  2, 43, 39, 45, 44,  1, 56, 46, 58, 47, 48, 49,
        50, 55, 51, 52, 54, 53]

 POC = [1, 2]

 Site ID = [60730001, 60731006, 60731008, 60731014, 60731016, 60731017, 60731022, 60731026]
 
 Dates range from 1/01/2020 to 12/31/2020